Quick_Deploy/vLLM/model_repository/vllm/1/model.py

```diff
@@ -0,0 +1,233 @@
+        """
+        Creates a task on the engine's event loop which is running on a separate thread.
+        """
+        assert (
+            self._shutdown_event.is_set() is False
+        ), "Cannot create tasks after shutdown has been requested"
+
+        return asyncio.run_coroutine_threadsafe(coro, self._loop)
+
+    def engine_loop(self, loop):
+        """
+        Runs the engine's event loop on a separate thread.
+        """
+        asyncio.set_event_loop(loop)
+        self._loop.run_until_complete(self.await_shutdown())
+
+    async def await_shutdown(self):
+        """
+        Primary coroutine running on the engine event loop. This coroutine is responsible for
+        keeping the engine alive until a shutdown is requested.
+        """
+        # first await the shutdown signal
+        while self._shutdown_event.is_set() is False:
+            await asyncio.sleep(5)
+
+        # Wait for the ongoing_requests
+        while self.ongoing_request_count > 0:
+            self.logger.log_info(
+                "Awaiting remaining {} requests".format(self.ongoing_request_count)
+            )
+            await asyncio.sleep(5)
+
+        self.logger.log_info("Shutdown complete")
+
+    def get_sampling_params_dict(self, params_json):
+        """
+        This functions parses the dictionary values into their
+        expected format.
+        """
+
+        dict = json.loads(params_json)
+
+        # Special parsing for the supported sampling parameters
+        # TODO: Add more parameters if needed
+        dict["temperature"] = float(dict["temperature"])
+        dict["top_p"] = float(dict["top_p"])
+
+        return dict
+
+    def postprocess(self, request_output):
+        """
+        Parses the output from the vLLM engine into text
+        response.
+        """
+        prompt = request_output.prompt
+        text_outputs = [
+            (prompt + output.text).encode("utf-8") for output in request_output.outputs
+        ]
+        return text_outputs
+
+    async def generate(self, request):
+        """
+        Forwards single request to LLM engine and returns responses.
+        """
+        response_sender = request.get_response_sender()
+        self.ongoing_request_count += 1
+        try:
+            request_id = random_uuid()
+            prompt = pb_utils.get_input_tensor_by_name(request, "PROMPT").as_numpy()[0]
+            stream = pb_utils.get_input_tensor_by_name(request, "STREAM").as_numpy()[0]
+            sampling_params_dict = self.get_sampling_params_dict(request.parameters())
+            sampling_params = SamplingParams(**sampling_params_dict)
+            results_generator = self.llm_engine.generate(
+                str(prompt), sampling_params, request_id
+            )
+
+            # Streaming case
+            async def stream_results() -> AsyncGenerator[bytes, None]:
+                async for request_output in results_generator:
+                    if self._shutdown_event.is_set():
```

Comment:
Here you are saying if the server is shutting down, then all current request should aborted.

I think a better behavior would be to raise an exception on any new requests coming into to the server, i.e. the client gets an error response with a message that the server is shutting down; however, all active requests still in flight are allowed to complete.

Does the `response_sender` which should be coupled to the state of the bidirectional grpc stream have an option to check if the stream was closed by the client?  If so, this is where you'd want to cancel the request with vllm and await the abort.

Reply:
> all active requests still in flight are allowed to complete.

Sure, we can revert to this behavior. We are working on propagating request cancellation from the gRPC streaming to the backends.